// @proofdict/textlint-rule-proofdict/lib/dictionary-storage.js
// MIT © 2018 azu
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.storage = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var _require = require("localstorage-ponyfill"),
    createLocalStorage = _require.createLocalStorage;

var Storage =
/*#__PURE__*/
function () {
  function Storage() {
    _classCallCheck(this, Storage);

    this.localStorage = createLocalStorage();
  }

  _createClass(Storage, [{
    key: "getItem",
    value: function getItem(name, defaultValue) {
      return this.localStorage.getItem(name, defaultValue);
    }
  }, {
    key: "setItem",
    value: function setItem(name, value) {
      return this.localStorage.setItem(name, value);
    }
  }, {
    key: "removeItem",
    value: function removeItem(name) {
      return this.localStorage.removeItem(name);
    }
  }]);

  return Storage;
}();

var storage = new Storage();
exports.storage = storage;
