// @textlint/linter-formatter/lib/types/src/Rule/TextlintRuleContext.js
// LICENSE : MIT
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TextlintRuleContextFixCommandGenerator_1 = require("./TextlintRuleContextFixCommandGenerator");
var TextlintRuleError_1 = require("./TextlintRuleError");
var TextlintRuleSeverityLevel_1 = require("./TextlintRuleSeverityLevel");
var assert = require("assert");
// instance for rule context
var ruleFixer = new TextlintRuleContextFixCommandGenerator_1.TextlintRuleContextFixCommandGenerator();
var TextlintRuleContext = /** @class */ (function () {
    function TextlintRuleContext(args) {
        var _this = this;
        /**
         * report function that is called in a rule
         */
        this.report = function (node, ruleError, _shouldNotUsed) {
            assert(!(node instanceof TextlintRuleError_1.TextlintRuleError), "1st argument should be node. Usage: `report(node, ruleError);`");
            assert(_shouldNotUsed === undefined, "3rd argument should not be used. Usage: `report(node, ruleError);`");
            if (ruleError instanceof TextlintRuleError_1.TextlintRuleError) {
                // severity come from `.textlintrc` option like `{ "<rule-name>" : { serverity: "warning" } } `
                _this._report({ ruleId: _this._ruleId, node: node, severity: _this._severityLevel, ruleError: ruleError });
            }
            else {
                var ruleReportedObject = ruleError;
                // severity come from report arguments like `report(node, { severity: 1 })`
                var level = ruleReportedObject.severity || TextlintRuleSeverityLevel_1.TextlintRuleSeverityLevel.error;
                _this._report({ ruleId: _this._ruleId, node: node, severity: level, ruleError: ruleReportedObject });
            }
        };
        /**
         * get file path current processing.
         */
        this.getFilePath = function () {
            return _this._sourceCode.getFilePath();
        };
        /**
         * Gets the source code for the given node.
         * @param {TxtNode=} node The AST node to get the text for.
         * @param {int=} beforeCount The number of characters before the node to retrieve.
         * @param {int=} afterCount The number of characters after the node to retrieve.
         * @returns {string} The text representing the AST node.
         */
        this.getSource = function (node, beforeCount, afterCount) {
            return _this._sourceCode.getSource(node, beforeCount, afterCount);
        };
        /**
         * get config base directory path
         * config base directory path often is the place of .textlintrc
         *
         * e.g.) /path/to/dir/.textlintrc
         * `getConfigBaseDir()` return `"/path/to/dir/"`.
         *
         * When using textlint as module, it is specified by `configBaseDir`
         * If not found the value, return undefined.
         *
         * You can use it for resolving relative path from config dir.
         * @returns {string|undefined}
         */
        this.getConfigBaseDir = function () {
            return _this._configBaseDir;
        };
        this._ruleId = args.ruleId;
        this._sourceCode = args.sourceCode;
        this._report = args.report;
        this._configBaseDir = args.configBaseDir;
        this._severityLevel = args.severityLevel;
        Object.freeze(this);
    }
    Object.defineProperty(TextlintRuleContext.prototype, "id", {
        /**
         * Rule id
         * @returns {string}
         */
        get: function () {
            return this._ruleId;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextlintRuleContext.prototype, "severity", {
        /**
         * severity level
         */
        get: function () {
            return this._severityLevel;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextlintRuleContext.prototype, "Syntax", {
        /**
         * Node's type values
         * @type {ASTNodeTypes}
         */
        get: function () {
            return this._sourceCode.getSyntax();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextlintRuleContext.prototype, "RuleError", {
        /**
         * CustomError object
         * @type {RuleError}
         */
        get: function () {
            return TextlintRuleError_1.TextlintRuleError;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextlintRuleContext.prototype, "fixer", {
        /**
         * Rule fixer command object
         * @type {RuleFixer}
         */
        get: function () {
            return ruleFixer;
        },
        enumerable: true,
        configurable: true
    });
    return TextlintRuleContext;
}());
exports.TextlintRuleContext = TextlintRuleContext;
